javascript/ql/src/Security/CWE-807/ConditionalBypass.ql

```diff
@@ -83,8 +83,25 @@ predicate isTaintedGuardForSensitiveAction(Sink sink, DataFlow::Node source, Sen
   )
 }
 
+/**
+ * Holds if `e` effectively guards access to `action` by returning or throwing early.
+ *
+ * Example: `if (e) return; action(x)`.
+ */
+predicate isEarlyAbortGuard(Sink e, SensitiveAction action) {
```

Comment:
This is a very syntactic check. Could we perhaps generalise this a bit using dominance and condition guard nodes?

Reply:
The syntactic check is quite deliberate as it is intended to whitelist a specific programming pattern.

The use of guard nodes and dominance is part of the Sink-character:
https://github.com/Semmle/ql/blob/16b29b2d08ea9232cad3ff0ab2d31724f66cfe23/javascript/ql/src/semmle/javascript/security/dataflow/ConditionalBypass.qll#L85-L90

Reply:
It seems to me that this check goes against the intention of the query. You are essentially saying that if it's possible to take a branch that doesn't reach `action`, then we whitelist that. But isn't the whole point of the query to flag cases where user input can cause `action` not to be executed?

Perhaps what we really want to exclude here are cases where avoiding `action` causes an abnormal termination of some sort. The proposed approach achieves that in the given example, but surely not in general?

I'm thinking of something like this:

```javascript
while (!verified) {
  if (user-controlled) {
    break;
  }
  verify();
}
```

Here, we would consider `verify()` to be a sensitive action, and the syntactic pattern looks much like in the example we want to whitelist, except that in this case we wouldn't want to whitelist.

As I am writing this, though, I am becoming less and less convinced that I actually understand what this query is meant to flag.

Reply:
Good point about `break`. I have reformulated the pattern to match `return` and `throw` only, see tests. 
As discussed elsewhere, I have also downgraded the precision of this query to `medium`.